library/std/src/io/copy.rs

```diff
@@ -54,14 +54,83 @@ where
     }
 }
 
-/// The general read-write-loop implementation of
-/// `io::copy` that is used when specializations are not available or not applicable.
-pub(crate) fn generic_copy<R: ?Sized, W: ?Sized>(reader: &mut R, writer: &mut W) -> io::Result<u64>
+/// The userspace read-write-loop implementation of `io::copy` that is used when
+/// OS-specific specializations for copy offloading are not available or not applicable.
+pub(crate) fn generic_copy<R: ?Sized, W: ?Sized>(reader: &mut R, writer: &mut W) -> Result<u64>
 where
     R: Read,
     W: Write,
 {
-    let mut buf = MaybeUninit::<[u8; super::DEFAULT_BUF_SIZE]>::uninit();
+    BufferedCopySpec::copy_to(reader, writer)
+}
+
+/// Specialization of the read-write loop that either uses a stack buffer
+/// or reuses the internal buffer of a BufWriter
+trait BufferedCopySpec: Write {
+    fn copy_to<R: Read + ?Sized>(reader: &mut R, writer: &mut Self) -> Result<u64>;
+}
+
+impl<W: Write + ?Sized> BufferedCopySpec for W {
+    default fn copy_to<R: Read + ?Sized>(reader: &mut R, writer: &mut Self) -> Result<u64> {
+        stack_buffer_copy(reader, writer)
+    }
+}
+
+impl<I: Write> BufferedCopySpec for BufWriter<I> {
+    fn copy_to<R: Read + ?Sized>(reader: &mut R, writer: &mut Self) -> Result<u64> {
+        if writer.capacity() < DEFAULT_BUF_SIZE {
+            return stack_buffer_copy(reader, writer);
+        }
+
+        let (mut len, buf, writer) = writer.as_raw_parts();
+
+        // FIXME: #42788
+        //
+        //   - This creates a (mut) reference to a slice of
+        //     _uninitialized_ integers, which is **undefined behavior**
+        //
+        //   - Only the standard library gets to soundly "ignore" this,
+        //     based on its privileged knowledge of unstable rustc
+        //     internals;
+        unsafe {
+            reader.initializer().initialize(MaybeUninit::slice_assume_init_mut(&mut buf[len..]));
+        }
+
+        let mut written = 0;
+        loop {
```

Comment:
It seems like this loop doesn't update the BufWriter's internal tracking of how much data is buffered, so use after copy returns will double write bits.

Rather than getting access to all of the raw bits of the writer's state, it seems like it would be cleaner to have APIs to get access to the unfilled part of the buffer, and to be able to write the buffer through (but not fully flush the inner writer) on demand.

Reply:
`as_raw_parts` sets the len to 0 ahead of time and it's the caller's responsibility to write out the data. The doc comment on `as_raw_parts` says as much. If that is unclear I can try to improve the comment.

Reply:
If the reader returns an error there will be dangling data in the output buffer, and this approach requires flushing the buffer out to the inner writer regardless of how much data is in it.

Reply:
To be more specific, I'm imagining roughly the inverse of BufRead, where there is a function to return the unfilled portion of the buffer, flushing if it happens to already be full to make space, and then a function to say that more data has been written into the buffer.

Reply:
I added a function that borrows the vec instead, which can be used for both.